pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt

```diff
@@ -1,35 +1,44 @@
 package com.pusher.pushnotifications.internal
 
 import com.pusher.pushnotifications.api.DeviceMetadata
-import com.squareup.moshi.adapters.PolymorphicJsonAdapterFactory
+import com.squareup.moshi.JsonClass
+import dev.zacsweers.moshisealed.annotations.TypeLabel
 import java.io.Serializable
 
-sealed class ServerSyncJob: Serializable {
-
-    companion object {
-        // the following is required for the Persistent Job Queue to read and write data correctly.
-        // if you add another ServerSyncJob, you must also add it below here too!
-        val jsonAdapterFactory: PolymorphicJsonAdapterFactory<ServerSyncJob>
-                = PolymorphicJsonAdapterFactory.of(ServerSyncJob::class.java, "ServerSyncJob")
-                .withSubtype(StartJob::class.java, "StartJob")
-                .withSubtype(RefreshTokenJob::class.java, "RefreshTokenJob")
-                .withSubtype(SubscribeJob::class.java, "SubscribeJob")
-                .withSubtype(UnsubscribeJob::class.java, "UnsubscribeJob")
-                .withSubtype(SetSubscriptionsJob::class.java, "SetSubscriptionsJob")
-                .withSubtype(ApplicationStartJob::class.java, "ApplicationStartJob")
-                .withSubtype(SetUserIdJob::class.java, "SetUserIdJob")
-                .withSubtype(StopJob::class.java, "StopJob")
-    }
-}
+@JsonClass(generateAdapter = true, generator = "sealed:ServerSyncJob")
+sealed class ServerSyncJob: Serializable { }
 
 // If you add new fields to any of these data classes, make sure they have
 // default values otherwise they can be null when reading the old stored
 // JSON from disk.
+@TypeLabel("StartJob")
+@JsonClass(generateAdapter = true)
 data class StartJob(val fcmToken: String, val knownPreviousClientIds: List<String>): ServerSyncJob()
```

Comment:
How about the migration? Won't we have a similar issue like the last time since we're changing how this sealed class hierarchy is unmarshalled? Won't there be any jobs persisted by a version prior to these changes of updating Moshi?

Reply:
Yeah. We'll have difficulties migrating if any of the underlying classes change. As we've not changed any of the structure of those classes everything seems to be happy! I've updated the comment at the top of `ServerSyncJob` to explain what will happen if those classes change, and explained that the `ServerSyncJobTests` are our single source of truth for what's expected to happen :-)

Reply:
Cool thanks for checking that!